### PR TITLE
doc: prefer for loops to for_each and try_for_each methods

### PR DESCRIPTION
**Pull request by mina86**

### CODE REVIEW

File: docs/style.md [hunk 1]
```diff
@@ -109,6 +109,44 @@ Note that for some *big* `Copy` types, notably `CryptoHash`, we sometimes use
...
+```
+
+**Rationale:** The `for_each` and `try_for_each` method don’t play nice with
+`break` and `continue` statements or with async IO (since `.await` is not
+possible).  And while `try_for_each` allows for the use of question mark
+operator, one may end up having to uses it twice: once inside the closure and
+second time outside the call to `try_for_each`.  Furthermore, usage of the
+functions often introduce some minor syntax noise.
+
+However, there are situations when those methods may lead to more readable code.
```
Comment (akhi3030) [inline docs/style.md:145]: Not part of this PR and a question for @matklad as well.  Every rule has an exception and in certain cases, if someone demonstrates the need to break one of these rules then they should be allowed to make exceptions.  Do we have some style agreed upon on how to do this?

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -2618,13 +2618,13 @@ mod test {
         let count_forwards_and_requests = |fixture: &ChunkTestFixture| -> (usize, usize) {
             let mut forwards_count = 0;
             let mut requests_count = 0;
-            fixture.mock_network.requests.read().unwrap().iter().for_each(|r| {
+            for r in fixture.mock_network.requests.read().unwrap().iter() {
                 match r.as_network_requests_ref() {
```
Comment (akhi3030) [inline chain/chunks/src/lib.rs:2622]: nit: although more tolerated in closures, can we move away from single letter variables please?  

```suggestion
            for request in fixture.mock_network.requests.read().unwrap().iter() {
                match requst.as_network_requests_ref() {
```

File: runtime/runtime-params-estimator/estimator-warehouse/src/check.rs [hunk 3]
```diff
@@ -100,7 +100,9 @@ pub(crate) fn create_report(db: &Db, config: &CheckConfig) -> anyhow::Result<Zul
         estimation_changes(db, &estimations, &commit_before, &commit_after, 0.1, config.metric)?;
 
     let mut report = ZulipReport::new(commit_before, commit_after);
-    warnings.into_iter().for_each(|w| report.add(w, Status::Warn));
+    for w in warnings {
+        report.add(w, Status::Warn)
```
Comment (akhi3030) [inline runtime/runtime-params-estimator/estimator-warehouse/src/check.rs:104]:
```suggestion
    for warning in warnings {
        report.add(warning, Status::Warn)
```

File: runtime/runtime/src/lib.rs [hunk 4]
```diff
@@ -1268,14 +1268,11 @@ impl Runtime {
...
             tracing::debug!(target: "runtime", node_counter = ?state_update.trie.get_trie_nodes_count());
-            result?.into_iter().try_for_each(
-                |outcome_with_id: ExecutionOutcomeWithId| -> Result<(), RuntimeError> {
-                    *total_gas_burnt =
-                        safe_add_gas(*total_gas_burnt, outcome_with_id.outcome.gas_burnt)?;
-                    outcomes.push(outcome_with_id);
-                    Ok(())
-                },
-            )?;
+            if let Some(outcome_with_id) = result? {
```
Comment (akhi3030) [inline runtime/runtime/src/lib.rs:1271]: Are we missing a loop here?

File: docs/style.md [hunk 5]
```diff
@@ -109,6 +109,44 @@ Note that for some *big* `Copy` types, notably `CryptoHash`, we sometimes use
 references for performance reasons. As a rule of thumb, `T` is considered *big* if
 `size_of::<T>() > 2 * size_of::<usize>()`.
 
+### Prefer for loops over `for_each` and `try_for_each` methods
```
Comment (nagisa) [inline docs/style.md:112]: I would love for this section to be more general and `for` vs `for_each` be just an example that the rule applies to. The same rationale applies to e.g. `{Option, Result}::and_then` as well, for example.

I was going to suggest something like:

> # Prefer native language constructs over methods taking a closure

but this wording doesn’t really work in cases like `map_err` which are much nicer with a method, and is probably way too general anyway. I wonder if there is any way to word this such that it only applies to the situations where a method would be a negative value…

File: runtime/runtime/src/lib.rs (hunk 4 above)
Comment (nagisa) [reply runtime/runtime/src/lib.rs:1271]: `Option`s are also iterators that output at most one value. In this case using an `if let` makes the code much more readable.

Comment (akhi3030) [reply runtime/runtime/src/lib.rs:1271]: Ah got it, thanks!

File: docs/style.md (hunk 1 above)
Comment (matklad) [reply docs/style.md:145]: I would say that styleguide is more of a lighthouse pointing the direction to go, rather than a set of rules that guard PRs. 

The following algo works:

**A** writes code with `.for_each`
**B** reviews and points to the style giude

Then, one of the three outcomes are possible. 

1. **A** changes to `for` because they don't care. 
2. **A** says "I aknowledge the styleguide, but in this particular cases there's this specific reason to prefer `.for_each`". `.for_each` stays without much arguming about specific reason
3. **A** says "but `.for_each` are more readable in general", in which case the discussion is moved out of scope for the particular PR. 



I think there should be a very low bar for making exceptions.

Comment (akhi3030) [reply docs/style.md:145]: I'm happy with this "process".

**Review by matklad — APPROVED** [review]

File: docs/style.md (hunk 5 above)
Comment (matklad) [reply docs/style.md:112]: I think there's also value in this being very specific: `.for_each` is the case where indeed this is very clear cut, generalizing might make this more obscure.

**Comment by mina86** [comment]
Closed by mistake.  Reopened in https://github.com/near/nearcore/pull/7255.

File: docs/style.md (hunk 5 above)
Comment (mina86) [reply docs/style.md:112]: Maybe we shouldn’t have this in style after all?  It is subjective after all; I wouldn’t advocate outright ban of `for_each`.  For example, I kinda prefer for_each version below (and I initially did not change it to a loop):

```
-        peers.iter().for_each(|p| update.delete::<schema::Peers>(p));
+        for p in peers {
+            update.delete::<schema::Peers>(p);
+        }
```

And regarding `and_then` (and also example in rust-analyzer has in its style guide), the below is much more readable than alternatives:

```rust
            Ok(accumulator)
                .and_then(|accumulator| safe_add_balance(accumulator, amount))
                .and_then(|accumulator| safe_add_balance(accumulator, locked))
                .map_err(|_| RuntimeError::UnexpectedIntegerOverflow)
```

Even though `Ok(accumulator)` is pointless in a sense, it improves readability.

Comment (matklad) [reply docs/style.md:112]: I do think it's very valuable to have some form of this advice in the style guide: there's bimodal distribution of how people write code, there are folks who do mostly imperative style, and folks who do mostly functional style. I think it's valuable if we have a general culture of "we are mostly imperative" or "we are mostly functional". 

There's specific problem that in OSS projects some contributors love submitting PR with highly high-order code, and its useful in this cases to point to the style guide and say "look, we appreciate the beauty of functional composition, but we explicitly prefer first-order code".

At the same time, this definitely should not be a "law" rule, more of the guideline, so some softening of the language in the styleguide would be good. As ra puts this, `This rule is more "soft" than others`. 

In this sense, I actually like being specific about `.for_each` -- it is the more clear-cut case, but still sets the overall "mostly imperative" culture. 

>For example, I kinda prefer for_each version below (and I initially did not change it to a loop):

For me, I would say that I also prefer the `.for_each` version in isolation, but, if I think about surrounding context (`for_each` is used rarely elsewhere) and evolution (the closure might get bigger, or it might get an `?`), this tips the scale in favor of `for` for me.

Comment (mina86) [reply docs/style.md:112]: OK, I dunno.  I’m just gonna merge this as is.  Someone smarter may expand the wording if they have an idea how.
